django/curator/tests/test_tag_deduplication.py

```diff
@@ -0,0 +1,117 @@
+        )
+        self.assertEqual(list, type(preprocessed_Tags))
+
+
+class TestTagClustering(TestCase):
+    def setUp(self):
+        for i in range(300):
+            id = TestTagClustering._id_generator()
+            CanonicalTag(name=id).save()
+            Tag(name=id + "1").save()
+            Tag(name=id + "2").save()
+            Tag(name=id + "3").save()
+
+    def test_uncertain_pairs(self):
+        tag_clustering = TagClustering()
+        uncertain_pairs = tag_clustering.uncertain_pairs()
+        self.assertEqual(list, type(uncertain_pairs))
+        self.assertEqual(1, len(uncertain_pairs))
+        self.assertEqual(2, len(uncertain_pairs[0]))
+        self.assertEqual(int, type(uncertain_pairs[0][0]["id"]))
+        self.assertEqual(str, type(uncertain_pairs[0][0]["name"]))
+        self.assertEqual(str, type(uncertain_pairs[0][0]["slug"]))
+
+    def test_cluster_tags(self):
+        clusters = self._cluster_tags()
+        self.assertEqual(list, type(clusters))
+
+        for cluster in clusters:
+            self.assertLess(0, len(cluster))
+            self.assertLess(0, len(cluster[0]))
+            self.assertLess(0, len(cluster[1]))
+
+    def _cluster_tags(self):
+        tag_clustering = TagClustering()
+        for i in range(300):
+            uncertain_pair = tag_clustering.uncertain_pairs()
+            tag_clustering.mark_pairs(uncertain_pair, i % 3 != 0)
+        return tag_clustering.cluster_tags()
+
+    def _id_generator(size=6):
+        return shortuuid.uuid()[:size]
+
+
+class TestTagGazetteering(TestCase):
+    def setUp(self):
+        for i in range(300):
+            _id = TestTagGazetteering._id_generator()
+            CanonicalTag(name=_id).save()
+            Tag(name=_id + "1").save()
+            Tag(name=_id + "2").save()
+            Tag(name=_id + "3").save()
+
+    def test_uncertain_pairs(self):
+        tag_clustering = TagGazetteering()
+        uncertain_pairs = tag_clustering.uncertain_pairs()
+        self.assertEqual(list, type(uncertain_pairs))
+        self.assertEqual(1, len(uncertain_pairs))
+        self.assertEqual(2, len(uncertain_pairs[0]))
+        self.assertEqual(int, type(uncertain_pairs[0][0]["id"]))
+        self.assertEqual(str, type(uncertain_pairs[0][0]["name"]))
+        self.assertEqual(str, type(uncertain_pairs[0][0]["slug"]))
+
+    def test_search(self):
+        clusters = self._search()
+        self.assertEqual(list, type(clusters))
+
+        for cluster in clusters:
+            self.assertEqual(int, type(cluster[0]))
+            self.assertEqual(tuple, type(cluster[1]))
+
+    def _search(self):
+        tag_clustering = TagGazetteering()
+        for i in range(300):
+            uncertain_pair = tag_clustering.uncertain_pairs()
+            tag_clustering.mark_pairs(uncertain_pair, i % 3 != 0)
+        return tag_clustering.search({1: {"id": 1, "name": "abms"}})
+
+    def _id_generator(size=6, chars=string.ascii_uppercase + string.digits):
```

Comment:
code duplication with the other `_id_generator` in `TagPreProcessingTestCase`, maybe we should move it up as a top-level function declaration used by both.